src/list-ext.satyg

```diff
@@ -43,27 +37,46 @@ module List : sig
   val init : 'a list -> ('a list) option
   val reverse-append : 'a list -> 'a list -> 'a list
   val reverse-map : ('a -> 'b) -> 'a list -> 'b list
+  val all : ('a -> bool) -> 'a list -> bool
+  val any : ('a -> bool) -> 'a list -> bool
+  val zip : 'a list -> 'b list -> ('a * 'b) list
+  val unzip : ('a * 'b) list -> 'a list * 'b list
+  val find : ('a -> bool) -> 'a list -> 'a option
+  val partition : ('a -> bool) -> 'a list -> 'a list * 'a list
+  val iterate : int -> ('a -> 'a) -> 'a -> 'a list
+  val repeat : int -> 'a -> 'a list
+  val cycle- : int -> 'a list -> 'a list
+  val acons : 'a -> 'b -> ('a * 'b) list -> ('a * 'b) list
+  val max : ('a Ord.t) implicit -> 'a list -> 'a option
+  val min : ('a Ord.t) implicit -> 'a list -> 'a option
+  val fold : 'b -> ('a -> 'b -> 'b) -> 'a list -> 'b
+
+  % unstable API (may subject to change in the future)
+  val fold-left : ('a -> 'b -> 'a) -> 'a -> 'b list -> 'a
+  val fold-lefti : (int -> 'a -> 'b -> 'a) -> 'a -> 'b list -> 'a
+  val fold-right : ('a -> 'b -> 'b) -> 'b -> 'a list -> 'b
+  val fold-left-adjacent : ('a -> 'b -> 'b option -> 'b option -> 'a) -> 'a -> 'b list -> 'a
+  val map-adjacent : ('a -> 'a option -> 'a option -> 'b) -> 'a list -> 'b list
+  val mapi-adjacent : (int -> 'a -> 'a option -> 'a option -> 'b) -> 'a list -> 'b list
```

Comment:
These six are unstable because the order of arguments is inconsistent.

Reply:
In what sense are they inconsistent?

Reply:
Well, a short answer is: for the first four, they are not consistent with List.fold. for the last two, I was not convinced.

A long answer:
Our coding/naming convention is a mix of Rust's and F#'s.
We are mainly following Rust's API, but unlike Rust we don't have methods and instead have partial application of function.
So in satysfi-base argument order is determined so that it is useful when combined with `|>` (this is the style used in F#).
For example, `Ref.set: 'a -> 'a ref -> unit` may look strange to OCaml users but it is quite natural in our style:
```
r |> Ref.set 1
```
This is like `r.set(1)` in a usual object-oriented language.
That is why I defined `List.fold` to be `'b -> ('a -> 'b -> 'b) -> 'a list -> 'b`, which is inconsistent with the first four of six.
```
one-to-five |> List.fold 0 (fun x y ->
  x * y
)
```
For the last two of six, `map-adjacent` and `mapi-adjacent`, I was not sure if the current argument order of the input closure is acceptable. Perhaps `'a option -> 'a -> 'a option -> 'b` could be better than `'a -> 'a option -> 'a option -> 'b`. But this is not so convincing. So I simply left them as unstable.

Reply:
BTW the reasons I adopted Rust's convention are:

1. It is a functional language.
2. Its API is large enough.
3. Its strength of type system is similar.

JaneStreet's Core was dropped due to point 3, as their API is based on the very strong type and module systems of OCaml. SATySFi lacks a trait (or type class) system like Rust, but it is emulatable.

Reply:
Got your idea, thank you for the clarification.